lib/Sema/CSSyntacticElement.cpp

```diff
@@ -905,6 +905,38 @@ class SyntacticElementConstraintGenerator
   }
 
   void visitBraceStmt(BraceStmt *braceStmt) {
+    auto &ctx = cs.getASTContext();
+
+    auto addResultDefault = [&](ClosureExpr *closure) {
+      if (context.getBody() == braceStmt) {
+        // If this closure has an empty body and no explicit result type
+        // let's bind result type to `Void` since that's the only type empty
+        // body can produce. Otherwise, if (multi-statement) closure doesn't
+        // have an explicit result (no `return` statements) let's default it to
+        // `Void`.
+        if (!constraints::hasExplicitResult(closure)) {
+          auto constraintKind =
+              (closure->hasEmptyBody() && !closure->hasExplicitResultType())
+                  ? ConstraintKind::Bind
+                  : ConstraintKind::Defaultable;
+
+          cs.addConstraint(constraintKind,
+                           cs.getClosureType(closure)->getResult(),
+                           ctx.TheEmptyTupleType,
+                           cs.getConstraintLocator(
+                               closure, ConstraintLocator::ClosureResult));
+        }
+      }
+    };
+
+    if (auto closure = cast<ClosureExpr>(
```

Comment:
Cannot use `cast` here because it’s not guaranteed that the context is indeed a closure, let’s add a method to context - `NullablePtr<ClosureExpr> getAsClosureExpr() const` and call it once at the beginning of this method and use the result everywhere else.

Reply:
Another issue with this check - it’s only applicable (the `participatesInInference` bit) when brace statement represents the whole body of the closure so we need `locator->directlyAt<ClosureExpr>()` here too.